api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,514 @@
+	}, nil
+}
+
+func (s MeshService) readLayer(layer *types.Layer, layerStatus pb.Layer_LayerStatus) (*pb.Layer, error) {
+	// Load all block data
+	var blocks []*pb.Block
+
+	// Save activations too
+	var activations []types.ATXID
+
+	for _, b := range layer.Blocks() {
+		txs, missing := s.Mesh.GetTransactions(b.TxIDs)
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
+		if len(missing) != 0 {
+			log.Error("could not find transactions %v from layer %v", missing, layer.Index())
+			return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+		}
+
+		activations = append(activations, b.ATXIDs...)
+
+		var pbTxs []*pb.Transaction
+		for _, t := range txs {
+			if t == nil {
+				log.Error("got empty transaction data from layer %v", layer.Index())
+				return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+			}
+			pbTxs = append(pbTxs, convertTransaction(t))
+		}
+		blocks = append(blocks, &pb.Block{
+			Id:           b.ID().Bytes(),
+			Transactions: pbTxs,
+		})
+	}
+
+	// Extract ATX data from block data
+	var pbActivations []*pb.Activation
+
+	// Add unique ATXIDs
+	atxs, matxs := s.Mesh.GetATXs(activations)
+	if len(matxs) != 0 {
+		log.Error("could not find activations %v from layer %v", matxs, layer.Index())
+		return nil, status.Errorf(codes.Internal, "error retrieving activations data")
+	}
+	for _, atx := range atxs {
+		pbatx, err := convertActivation(atx)
+		if err != nil {
+			log.Error("error serializing activation data: %s", err)
+			return nil, status.Errorf(codes.Internal, "error serializing activation data")
+		}
+		pbActivations = append(pbActivations, pbatx)
+	}
+
+	stateRoot, err := s.Mesh.GetLayerStateRoot(layer.Index())
+	if err != nil {
+		log.Error("error getting state root for layer %v", layer.Index())
+		return nil, status.Errorf(codes.Internal, "error getting layer data")
+	}
+	return &pb.Layer{
+		Number:        layer.Index().Uint64(),
+		Status:        layerStatus,
+		Hash:          layer.Hash().Bytes(),
+		Blocks:        blocks,
+		Activations:   pbActivations,
+		RootStateHash: stateRoot.Bytes(),
+	}, nil
+}
+
+// LayersQuery returns all mesh data, layer by layer
+func (s MeshService) LayersQuery(ctx context.Context, in *pb.LayersQueryRequest) (*pb.LayersQueryResponse, error) {
+	log.Info("GRPC MeshService.LayersQuery")
+
+	// Get the latest layers that passed both consensus engines.
+	lastLayerPassedHare := s.Mesh.LatestLayerInState()
+	lastLayerPassedTortoise := s.Mesh.ProcessedLayer()
+
+	layers := []*pb.Layer{}
+	for l := uint64(in.StartLayer); l <= uint64(in.EndLayer); l++ {
+		layerStatus := pb.Layer_LAYER_STATUS_UNSPECIFIED
+
+		// First check if the layer passed the Hare, then check if it passed the Tortoise.
+		// It may be either, or both, but Tortoise always takes precedence.
+		if l <= lastLayerPassedHare.Uint64() {
+			layerStatus = pb.Layer_LAYER_STATUS_APPROVED
+		}
+		if l <= lastLayerPassedTortoise.Uint64() {
```

Comment:
won't this cause all lauers to have a confirmed status - evet if tortoise haven't confirmed them yet?

Reply:
Hmm. I reread this logic and I think it's correct. If a layer has passed neither Tortoise nor Hare, it's marked UNSPECIFIED. If it's passed only Hare, it's marked APPROVED. If it's passed only Tortoise, or both, it's marked CONFIRMED. Is there an error in this logic?

Please make sure you double-check lines 348-349 above, too.

Reply:
BTW the logic for this was determined here: https://github.com/spacemeshos/api/issues/89#issuecomment-655625229

Reply:
I meant that `s.Mesh.ProcessedLayer()` is the layer last pushed to tortoise, the tortoise then outputs newPbase which is the actual layer that was confirmed

Reply:
https://github.com/spacemeshos/go-spacemesh/issues/2138